trace_decoder/src/wire.rs

```diff
@@ -0,0 +1,317 @@
+        struct EncodeKeyFlags: u8 {
+            const ODD = 0b0000_0001;
+            const TERMINATED = 0b0000_0010;
+        }
+    }
+    let v = match bytes.split_first() {
+        // BUG(spec): the previous implementation said that Erigon does this
+        (only, &[]) => nunny::vec![U4::new(*only).ok_or(Error("excess bits in single nibble"))?],
+        (flags, rest) => {
+            // check the flags
+            let flags = EncodeKeyFlags::from_bits(*flags)
+                .ok_or(Error("unrecognised bits in flags for key encoding"))?;
+            // BUG(spec)?: the previous implementation ignored this flag, and
+            //             our tests fail without it - perhaps it's &-ed with
+            //             the prior U4?
+            // if flags.contains(EncodeKeyFlags::TERMINATED) {
+            //     match rest.split_last() {
+            //         Some((0x10, new_rest)) => rest = new_rest,
+            //         _ => return Err(Error("bad terminator for key")),
+            //     }
+            // }
+
+            // treat the final byte as special...
+            let (last, rest) = match rest.split_last() {
+                Some((last, rest)) => (Some(*last), rest),
+                None => (None, &[][..]),
+            };
+
+            // ...according to the ODD flag
+            let tail = match (last, flags.contains(EncodeKeyFlags::ODD)) {
+                (None, true) => return Err(Error("span was zero but odd flag was set")),
+                (None, false) => Either::Left(iter::empty::<U4>()),
+                (Some(left), true) => Either::Right(Either::Left(iter::once(
+                    U4::new(left.rotate_right(4)).ok_or(Error("unrecognised bits in nibble"))?,
+                ))),
+                (Some(packed), false) => {
+                    let both = U4x2 { packed };
+                    Either::Right(Either::Right([both.left(), both.right()].into_iter()))
+                }
+            };
+
+            // parse the rest of the bytes as nibbles
+            nunny::Vec::new(
+                rest.iter()
+                    .copied()
+                    .flat_map(|packed| {
+                        let both = U4x2 { packed };
+                        [both.left(), both.right()]
+                    })
+                    .chain(tail)
+                    .collect(),
+            )
+            .expect("an empty `rest` must be caught by the Erigon special case")
+        }
+    };
+    Ok(v)
+}
+
+fn array<const N: usize>(input: &mut &[u8]) -> PResult<[u8; N]> {
+    take(N)
+        .map(|it: &[u8]| it.try_into().expect("take has already selected N bytes"))
+        .parse_next(input)
+}
+
+#[cfg(test)]
+#[track_caller]
+fn do_test<'a, T: PartialEq + core::fmt::Debug>(
+    src: &'a [u8],
+    expected: T,
+    mut parser: impl winnow::Parser<&'a [u8], T, winnow::error::ContextError>,
+) {
+    let actual = parser.parse(src.as_ref()).unwrap();
+    assert_eq!(expected, actual)
+}
+
```

Comment:
Non blocking: should we add extensive testing here to cover [specification](https://github.com/0xPolygonHermez/cdk-erigon/blob/zkevm/turbo/trie/WITNESS.md)

Reply:
#362